src/hotspot/share/oops/klass.cpp

```diff
@@ -163,20 +163,41 @@ void Klass::release_C_heap_structures(bool release_constant_pool) {
   if (_name != nullptr) _name->decrement_refcount();
 }
 
-bool Klass::search_secondary_supers(Klass* k) const {
-  // Put some extra logic here out-of-line, before the search proper.
-  // This cuts down the size of the inline method.
-
-  // This is necessary, since I am never in my own secondary_super list.
-  if (this == k)
-    return true;
+bool Klass::linear_search_secondary_supers(Klass* k) const {
   // Scan the array-of-objects for a match
+  // FIXME: We could do something smarter here, maybe a vectorized
+  // comparison or a binary search, but is that worth any added
+  // complexity?
   int cnt = secondary_supers()->length();
   for (int i = 0; i < cnt; i++) {
     if (secondary_supers()->at(i) == k) {
-      ((Klass*)this)->set_secondary_super_cache(k);
+      if (UseSecondarySupersCache) {
+        ((Klass*)this)->set_secondary_super_cache(k);
```

Comment:
I kinda hate this because we're casting away `const`, which is UB. I think I'd just take it out, but once I do that, I don't think anything sets `_secondary_super_cache`.

Reply:
IMO it's OK if C++ runtime omits `_secondary_super_cache` accesses irrespective of whether `UseSecondarySupersCache` is set or not. I'm fine with addressing it separately.